arroyo-state/src/lib.rs

```diff
@@ -115,10 +130,30 @@ pub trait BackingStore {
         value: &mut V,
     );
 
+    async fn delete_data_triple<K: Key>(
+        &mut self,
+        table: char,
+        table_type: TableType,
+        timestamp: SystemTime,
+        key: &mut K,
+    );
+
+    async fn delete_data_triples_before<K: Key>(
+        &mut self,
+        table: char,
+        table_type: TableType,
+        before: SystemTime,
+        key: &mut K,
+    );
+
     async fn write_key_value<K: Key, V: Data>(&mut self, table: char, key: &mut K, value: &mut V);
+    async fn delete_key_value<K: Key>(&mut self, table: char, key: &mut K);
 
-    async fn get_global_key_values<K: Key, V: Data>(&self, table: char) -> Vec<(K, V)>;
-    async fn get_key_values<K: Key, V: Data>(&self, table: char) -> Vec<(K, V)>;
```

Comment:
I combined these functions since the only difference was the key range.

Reply:
I don't love this, as it pushes up a single key assignment strategy (one contiguous hash range per subtask) into the API. Having the two methods results in semantics of "get the data the subtask owns" and "get all of the data for the subtask".

Reply:
Ok I've added the two functions back to the trait but made their implementations both use the same function.